### PR TITLE
Add workers() option to the Redis destination

### PR DESCRIPTION
**Pull request by OverOrion**
This PR adds `workers()` support for the `redis()` destination driver.

This increases the throughput of the Redis destination driver.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by alltilla** [comment]
@kira-syslogng ok to test

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
@kira-syslogng add to whitelist

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/redis/redis-worker.c [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+LogThreadedDestWorker *redis_dw_new(LogThreadedDestDriver *owner, gint worker_index)
+{
+  RedisDestWorker *self = g_new0(RedisDestWorker, 1);
+
+  log_threaded_dest_worker_init_instance(&self->super, owner, worker_index);
+
+  self->super.thread_init = redis_worker_thread_init;
+  self->super.thread_deinit = redis_worker_thread_deinit;
+  self->super.connect = redis_dd_connect;
+  self->super.disconnect = redis_dd_disconnect;
```
Comment (Kokan) [inline modules/redis/redis-worker.c:303]: The `dd` usually refers to destination driver, and the `dw` a destination worker. Please use the proper naming, in this case `redis_dw_disconnect` or the `redis_worker_disconnect` is fine. But use only either (`worker`/`dw`) for all method in this compilation unit.

File: modules/redis/redis-worker.c [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+      append_unsafe_utf8_as_escaped_text(full_command, self->argv[i], self->argvlen[i], "\"");
+      g_string_append(full_command, "\"");
+    }
+  return full_command->str;
+}
+
+
+
+/*
+ * Worker thread
```
Comment (Kokan) [inline modules/redis/redis-worker.c:96]: Same as the other worker thread comment.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

File: modules/redis/redis-worker.c [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+static gboolean
+check_connection_to_redis(RedisDestWorker *self)
+{
+  return send_redis_command(self, "ping");
+}
+
+static gboolean
+authenticate_to_redis(RedisDestWorker *self, const gchar *password)
+{
+  return send_redis_command(self, "AUTH %s", ((RedisDriver *) self->super.owner)->auth);
```
Comment (MrAnno) [inline modules/redis/redis-worker.c:224]: The `password` parameter can be used here.

File: modules/redis/redis-worker.c [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+#include "scratch-buffers.h"
+#include "utf8utils.h"
+
+
+
+static inline void
+_fill_template(RedisDestWorker *self, LogMessage *msg, LogTemplate *template, gchar **str,
+               gsize *size)
+{
+  RedisDestWorker *s = (RedisDestWorker *) self;
```
Comment (MrAnno) [inline modules/redis/redis-worker.c:38]: `self` has type `RedisDestWorker`, this cast is not necessary.

File: modules/redis/redis-worker.c [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+      log_template_format(template, msg, &options, buffer);
+      *size = buffer->len;
+      *str = buffer->str;
+    }
+}
+
+static void
+_fill_argv_from_template_list(RedisDestWorker *s, LogMessage *msg)
+{
+  RedisDestWorker *self = (RedisDestWorker *) s;
```
Comment (MrAnno) [inline modules/redis/redis-worker.c:62]: This cast is not necessary.

**Review by MrAnno — DISMISSED** [review]
Perfect, thanks!

**Comment by MrAnno** [comment]
#3741 has been merged.
@OverOrion Could you please rebase and follow the changes up in the grammar? (batching and workers-related non-terminal symbols)

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by Kokan** [comment]
The `workers()` option does not work, you have to opt in in the grammar using `threaded_dest_driver_workers_option`.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

**Review by MrAnno — APPROVED** [review]

**Comment by MrAnno** [comment]
Thanks.

**Pull request merged**